scripts/addr2line.py

```diff
@@ -83,41 +84,148 @@ def __call__(self, address):
         self._input.stdin.flush()
         return self._read_resolved_address()
 
+class KernelResolver:
+    """A reserver for kernel addresses which tries to read from /proc/kallsyms."""
+
+    LAST_SYMBOL_MAX_SIZE = 1024
+
+    def __init__(self, binary, *args, **kwargs):
+        assert binary == '<kernel>', 'this resolver only makes sense for the kernel'
+        syms : list[tuple[int, str]] = []
+
+        ksym_re = re.compile(r'(?P<addr>[0-9a-f]+) (?P<type>.+) (?P<name>\S+)')
+
+        warnings_left = 10
+        try:
+            f = open('/proc/kallsyms', 'r')
+        except OSError as e:
+            self.error = f'Cannot open /proc/kallsyms: {e}'
+            print(self.error)
+            return
+
+        try:
+            for line in f:
+                m = ksym_re.match(line)
+                if not m:
+                    if warnings_left > 0: # don't spam too much
+                        print(f'WARNING: /proc/kallsyms regex match failure: {line.strip()}', file=sys.stdout)
+                        warnings_left =- 1
+                else:
+                    syms.append((int(m.group('addr'), 16), m.group('name')))
+        finally:
+            f.close()
+
+        if not syms:
+            # make empty kallsyms (?) an error so we can assum len >= 1 below
+            self.error = 'kallsyms was empty'
+            print(self.error)
+            return
+
+        syms.sort()
+
+        if syms[-1][0] == 0:
+            # zero values for all symbols means that kptr_restrict blocked you
+            # from seeing the kernel symbol addresses
+            print('kallsyms is restricted, set /proc/sys/kernel/kptr_restrict to 0 to decode')
+            self.error = 'kallsyms is restricted'
+            return
+
+        self.error = None
+        # split because bisect can't take a key func before 3.10
+        self.sym_addrs : tuple[int]
+        self.sym_names : tuple[str]
```

Comment:
zip doesn't seem to propagate though the types correct, e.g., `zip( (1, 'x'), (1, 'x') )` results in tuples `(1, 1)` and `('x', 'x')` which are compatible with types `tuple[int]` and `tuple[str]`, but they both end up as `tuple[int | str]`. 

This is almost certainly because this use of `zip` is in the opposite direction as the usual one (TIL: `zip` is its own inverse, i.e. you don't need a "reverse zip" operation).